crates/rethnet_evm/src/mempool.rs

```diff
@@ -148,21 +148,30 @@ impl MemPool {
             sender: &AccountInfo,
         ) -> bool {
             U256::from(transaction.gas_limit()) <= *block_gas_limit
-                && transaction.nonce() >= sender.nonce
                 && transaction.upfront_cost() <= sender.balance
         }
 
         for entry in self.pending_transactions.iter_mut() {
             let (caller, transactions) = entry;
             let sender = state.basic(*caller)?.unwrap_or_default();
 
+            // Remove all finalized transactions
+            transactions.retain(|transaction| {
+                let should_retain = transaction.nonce() >= sender.nonce;
+
+                if !should_retain {
+                    self.hash_to_transaction.remove(transaction.hash());
+                }
+
+                should_retain
+            });
+
             if let Some((idx, _)) = transactions
                 .iter()
                 .enumerate()
                 .find(|(_, transaction)| !is_valid_tx(transaction, &self.block_gas_limit, &sender))
             {
-                // Question: Do we ever need to consider tx.nonce < sender.nonce, due to manual modifications?
-
+                // Move all consequent transactions to the future queue
                 let mut invalidated_transactions = transactions.split_off(idx);
```

Comment:
Tbh I don't understand why we can split off at that index. Why are all the transactions after that are guaranteed to be invalid?

Reply:
Transactions are sorted by insertion order and transaction can only be inserted when it's an increment of one nonce. Thus, if we remove one transaction of an author, we can also removed all consequent ones.